Add tests to check undo not resetting tools

ClFeSc opened the pull request:
Closes #38 
Closes #69 

This currently adds some tests checking for tools changing when using the unfixed undo. As far as I'm concerned we could remove them when we are ready with the other parts of this PR. They currently exist to check that the tests actually tests something that could be a problem. The same applies to the changes in the `M2Backend` class (except for the assignment of `currentTool` in `saveStateInCurrentUserAction`).

kpostnov commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/brushSizeTestUsingUndo.testing..st:
```diff
@@ -0,0 +1,26 @@
+	self compareFormWith: #brushSizeTest.
+	self backend setBrushAndEraserSizeTo: 40.
+	self drawAt: 20 asPoint.
+	"TODO: Add assert that this actually changed something."
```
Remove comment.

ClFeSc replied:
Yes, but only after we found the time to create a comparison form for this case. This test would currently pass when the `drawAt:` call didn't change anything, ignoring everything `undo` could do wrong.

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/multipleFillAreaTestUsingUndo.testing..st:
```diff
@@ -0,0 +1,30 @@
+		setBrushAndEraserSizeTo: 50;
+		drawCircle.
+	self drawAt: 50 asPoint.
+	self compareFormWith: #brushSizeTest.
```
```suggestion
	self
		drawAt: 50 asPoint;
		compareFormWith: #brushSizeTest.
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/multipleFillAreaTestUsingUndo.testing..st:
```diff
@@ -0,0 +1,30 @@
+		fillShape;
+		penColor: Color white.
+	self drawAt: 50 asPoint.
+	self compareFormWith: #whiteBackground.
```
```suggestion
	self
		drawAt: 50 asPoint;
		compareFormWith: #whiteBackground.
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/simpleTextUsingUndo.testing..st:
```diff
@@ -0,0 +1,20 @@
+		fillShape;
+		penColor: Color blue.
+	self drawAt: 10 asPoint.
+	self compareFormWith: #blueBackground.
```
```suggestion
	self
		drawAt: 10 asPoint;
		compareFormWith: #blueBackground.
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/simpleTextUsingUndo.testing..st:
```diff
@@ -0,0 +1,20 @@
+	self compareFormWith: #blueBackground.
+	self backend penColor: Color white.
+	self drawAt: 10 asPoint.
+	self compareFormWith: #whiteBackground.
```
```suggestion
	self
		drawAt: 10 asPoint;
		compareFormWith: #whiteBackground.
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testBrushSize.st:
```diff
@@ -2,6 +2,6 @@ testing
-		clickAt: 50 asPoint.
+		drawCircle.
+	self drawAt: 50 asPoint.
 	self compareFormWith: #brushSizeTest
```
```suggestion
	self
		drawAt: 50 asPoint;
		compareFormWith: #brushSizeTest
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testCircleBrush.st:
```diff
@@ -1,6 +1,5 @@
\ No newline at end of file
+	self backend drawCircle.
+	self drawAt: 20 asPoint.
+	self compareFormWith: #circleBrushTest
```
```suggestion
	self
		drawAt: 20 asPoint;
		compareFormWith: #circleBrushTest
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testEraserSize.st:
```diff
@@ -2,8 +2,8 @@ testing
+	self drawAt: 75@75.
+	self backend drawEraser.
+	self drawAt: 75@75.
 	self compareFormWith: #whiteBackground
```
```suggestion
	self
		drawAt: 75 asPoint;
		compareFormWith: #whiteBackground
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testEraserSize.st:
```diff
@@ -2,8 +2,8 @@ testing
-		drawEraser;
-		clickAt: 75@75.
+		drawCircle.
+	self drawAt: 75@75.
```
```suggestion
	self drawAt: 75 asPoint.
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testSimpleText.st:
```diff
@@ -1,8 +1,7 @@
+	self drawAt: 25 asPoint.
 	self backend temporaryTextMorph contents: 'test'.
-	self backend clickAt: 20 asPoint.
+	self drawAt: 20 asPoint.
 	self compareFormWith: #simpleText
```
```suggestion
	self
		drawAt: 20 asPoint;
		compareFormWith: #simpleText
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testTriangleBrush.st:
```diff
@@ -1,6 +1,5 @@
-	self compareFormWith: #triangleBrushTest
\ No newline at end of file
+	self backend drawTriangle.
+	self drawAt: 20 asPoint.
+	self compareFormWith: #triangleBrushTest
```
```suggestion
	self
		drawAt: 20 asPoint;
		compareFormWith: #triangleBrushTest
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testUndoBrushSize.st:
```diff
@@ -0,0 +1,9 @@
+	self backend
+		setBrushAndEraserSizeTo: 50;
+		drawCircle.
+	self drawAt: 50 asPoint.
+	self compareFormWith: #brushSizeTest.
```
```suggestion
	self
		drawAt: 50 asPoint;
		compareFormWith: #brushSizeTest.
```

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testUndoSimpleText.st:
```diff
@@ -1,10 +1,9 @@
+	self drawAt: 25 asPoint.
 	self backend temporaryTextMorph contents: 'test'.
-	self backend clickAt: 20 asPoint.
+	self drawAt: 20 asPoint.
 	self compareFormWith: #simpleText.
```
```suggestion
	self
		drawAt: 20 asPoint;
		compareFormWith: #simpleText.
```

DevSchmidtchen commented on packages/SketchMorph2-Core.package/M2Backend.class/instance/unfixedUndo.st:
```diff
@@ -0,0 +1,3 @@
+tool selection
+unfixedUndo
```
Why do we need to test unfixed implementation?

DevSchmidtchen commented on packages/SketchMorph2-Core.package/M2Backend.class/instance/undoBlock.st:
```diff
@@ -0,0 +1,7 @@
+tool selection
+undoBlock
```
Should we name a method based on the return type?

DevSchmidtchen replied:
Maybe it could be called `undoActions` or anything without using the word `block`

DevSchmidtchen commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/multipleFillAreaTestUsingUndo.testing..st:
```diff
@@ -0,0 +1,32 @@
+testblocks
+multipleFillAreaTestUsingUndo: anUndoBlock testing: aTestBlock
```
What is the reason for such a complex test which includes tests for things not related to undo itself (like brush tool, fill area tool etc.). I think the test should be as small as possible to only test one specific detail.

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/simpleTextUsingUndo.testing..st:
```diff
@@ -0,0 +1,22 @@
+simpleTextUsingUndo: anUndoBlock testing: aTestBlock
+	self backend
+		fillShape;
+		penColor: Color blue.
+	self
+		drawAt: 10 asPoint;
+		compareFormWith: #blueBackground.
```
Isn't that unnecessary?

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testUndoBrushSize.st:
```diff
@@ -0,0 +1,10 @@
+testUndoBrushSize
+	self backend
+		setBrushAndEraserSizeTo: 50;
+		drawCircle.
+	self
+		drawAt: 50 asPoint;
+		compareFormWith: #brushSizeTest.
```
I don't think we should test that here, since we have a dedicated test for it.

maximilian-franz commented on packages/SketchMorph2-Tests.package/M2BackendTests.class/instance/testUndoSimpleText.st:
```diff
@@ -0,0 +1,10 @@
+testUndoSimpleText
+	self backend addText.
+	self drawAt: 25 asPoint.
+	self backend temporaryTextMorph contents: 'test'.
+	self
+		drawAt: 20 asPoint;
+		compareFormWith: #simpleText.
```
I don't think we should test that here, since we have a dedicated test for it.

ClFeSc commented:
I have changed the undo tests to check whether undoing an action always
- results in the image it was before applying that action, and
- preserves the whole state of the context.

This means that tests testing unfixed undo not working have been removed because there is no point in testing something that is expecting to not work. If it works now, we're happy, regardless of what was before. I also removed the comparisons with fixed images when undoing since we want to test that undo restores the state prior to the user action and not that undo resets it to some magic state that is defined by the user actions that are tested elsewhere.

We can now decide to add more tests for undo by simply adding a test method calling `execute:` with no further need of setup.

DevSchmidtchen commented on packages/SketchMorph2-Tests.package/M2DrawingComponentTests.class/instance/execute..st:
```diff
@@ -0,0 +1,14 @@
+support
+execute: aBlockArray
```
`execute` doesn't imply testing the undo functionality here. Maybe the name should include the purpose of testing undo.

ClFeSc replied:
Yeah, you're right. This method was once designated to have more parameters, one of which would have implied the undo action. Since they are not needed anymore, I'll rename the method.

DevSchmidtchen commented on packages/SketchMorph2-Core.package/M2Backend.class/instance/undoBlock.st:
```diff
@@ -0,0 +1,7 @@
+tool selection
+undoBlock
```
Still think this name should be changed. What do you think?
